### ISSUE TITLE
Why `isVar f` here?

### ISSUE THREAD

**Issue by josephcsible** [ndmitchell/hlint#822]
https://github.com/ndmitchell/hlint/blob/1e22e318fa40411d7782d391a7d9396d9fc72675/data/hlint.yaml#L173

I assume `isVar f` is here for a reason, but I can't figure out what that reason is.

**Comment by ndmitchell** [comment]
Because if you have a more complex expression it can look a bit ugly. e.g.

```
map (f a b c) (zip x y)
```

Seems not as bad as:

```
zipWith (curry (f a b c)) x y
```

It's just to try and not trigger a fairly complex hint (`curry` isn't all that common) unless it's very simple outcome.

**Comment by josephcsible** [comment]
I was thinking about the case `map (\(x,y) -> foobar) (zip xs ys)`. I'd like it to be rewritten to `zipWith (\x y -> foobar) xs ys`. To do that, while avoiding the rewrite you mentioned, what if we changed `isVar f` to `not (isApp f)` in that rule, and then added these two rules?

```yaml
- warn: {lhs: "uncurry (\\x y -> z)", rhs: "\\(x,y) -> z"}
- warn: {lhs: "curry (\\(x,y) -> z)", rhs: "\\x y -> z"}
```

I can't think of any non-contrived cases of where `map` would get something that's not a `Var`, `App`, or lambda, and whenever it's a lambda, the new rules would remove the curry.

**Comment by ndmitchell** [comment]
Makes a lot of sense. The basic rationale was that you didn't want to introduce more brackets. To a first approximation everything either has brackets, or doesn't need them. The one exception is app, so it makes sense from that perspective.
